storage/grpc_writer.go

```diff
@@ -291,6 +293,9 @@ func (s *gRPCAppendBidiWriteBufferSender) sendOnConnectedStream(buf []byte, offs
 			if resp.GetResource() != nil {
 				obj = resp.GetResource()
 			}
+			if s.obj != nil && resp.GetPersistedSize() > 0 {
+				s.obj.Size = resp.GetPersistedSize()
+			}
```

Comment:
So s.obj is only set by the first response, except for `size`? And is only used by the constructor? I am confused by this flow a bit, maybe comments would help

Reply:
I renamed and added some comments. Basically the object resource is returned 1. on the initiation of the first write. 2. on takeover. 3. on finalization. In previous write flows we were relying (1) or (3) to populate the object; but for takeover + unfinalized we need to populate it on takeover and also update the size at the end to reflect the actual persisted size.